### PR TITLE
Add support for running the agent behind a web proxy (--proxyurl)

### PR DESCRIPTION
**Pull request by psdgal**
Hi,

could you please integrate this changes in order to support agents running behind a web proxy?

Thank you.

### CODE REVIEW

File: VSTSAgent/VSTSAgent.psm1 [hunk 1]
```diff
@@ -243,6 +246,10 @@ function Install-VSTSAgent {
         [parameter(Mandatory = $true)]
         [uri]$ServerUrl,
 
+        [parameter(Mandatory = $false)]
+        [ValidateScript({ $_ -match "http://\d{1,3}[.]\d{1,3}[.]\d{1,3}[.]\d{1,3}:\d{1,5}" })]
```
Comment (jwittner) [inline VSTSAgent/VSTSAgent.psm1:250]: Can we guarantee all proxy urls will match this? E.g. what about https urls? This looks like it forces an IP to be used as well? Maybe just skip the validation?

File: VSTSAgent/DSCResources/xVSTSAgent/xVSTSAgent.psm1 [hunk 2]
```diff
@@ -171,6 +175,7 @@ function Set-TargetResource {
 
         if ( $Work ) { $installArgs['Work'] = $Work }
         if ( $LogonCredential ) { $installArgs['LogonCredential'] = $LogonCredential }
+        if ( $ProxyUrl ) { $installArgs['ProxyUrl'] = $ProxyUrl }
```
Comment (jwittner) [inline VSTSAgent/DSCResources/xVSTSAgent/xVSTSAgent.psm1:178]: Would be really nice to pass through the proxy username/password args as well.

**Review by jwittner — CHANGES_REQUESTED** [review]

File: VSTSAgent/DSCResources/xVSTSAgent/xVSTSAgent.psm1 (hunk 2 above)
Comment (psdgal) [reply VSTSAgent/DSCResources/xVSTSAgent/xVSTSAgent.psm1:178]: Unfortunately I have no means to test the combination: url + credential + password.

File: VSTSAgent/DSCResources/xVSTSAgent/xVSTSAgent.psm1 [hunk 3]
```diff
@@ -131,6 +131,10 @@ function Set-TargetResource {
         [System.String]
         $Work,
 
+        [parameter(Mandatory = $false)]
+        [System.String]
+        [string]$ProxyUrl,
```
Comment (jwittner) [inline VSTSAgent/DSCResources/xVSTSAgent/xVSTSAgent.psm1:136]: Sorry for missing this the first time, but we also need to add this flag to the Test-TargetResource function in this file (and we should test that the proxy setup is the same if we do find an agent, if possible). We also need to add this data to the .mof file next to these files.

**Review by jwittner — CHANGES_REQUESTED** [review]

File: VSTSAgent/DSCResources/xVSTSAgent/xVSTSAgent.psm1 (hunk 3 above)
Comment (psdgal) [reply VSTSAgent/DSCResources/xVSTSAgent/xVSTSAgent.psm1]: Done :)

File: VSTSAgent/VSTSAgent.psm1 [hunk 4]
```diff
@@ -271,6 +277,7 @@ function Install-VSTSAgent {
     if ( $MinimumVersion ) { $findArgs['MinimumVersion'] = $MinimumVersion }
     if ( $MaximumVersion ) { $findArgs['MaximumVersion'] = $MaximumVersion }
     if ( $RequiredVersion ) { $findArgs['RequiredVersion'] = $RequiredVersion }
+    if ( $ProxyUrl ) {  $findArgs['ProxyUrl'] = $ProxyUrl }
```
Comment (jwittner) [inline VSTSAgent/VSTSAgent.psm1:280]: I don't think you want this here as Find-VSTSAgent doesn't support that arg and I think this will throw at runtime.

Comment (jwittner) [reply VSTSAgent/VSTSAgent.psm1:280]: At this point we're trying to find the agent to install.

**Review by jwittner — CHANGES_REQUESTED** [review]
Just one small change to undo and I think this looks great!

File: VSTSAgent/VSTSAgent.psm1 (hunk 4 above)
Comment (psdgal) [reply VSTSAgent/VSTSAgent.psm1]: Should be ok now :)

**Comment by psdgal** [comment]
Any news regarding that?

**Comment by jwittner** [comment]
Thanks for the changes!

**Review by jwittner — APPROVED** [review]

**Pull request merged**